test/lib/study_cleanup_tools_test.rb

```diff
@@ -0,0 +1,105 @@
+require 'test_helper'
+
+# test all validations to ensure security is being enforced that will block accidental deletes of user studies/workspaces
+# does not directly test cleanup methods as these would destroy seed data needed for other tests
+class StudyCleanupToolsTest < ActiveSupport::TestCase
+
+  include Minitest::Hooks
+  include SelfCleaningSuite
+  include TestInstrumentor
+
+  before(:all) do
+    @user = FactoryBot.create(:user, test_array: @@users_to_clean)
+    @basic_study = FactoryBot.create(:study, name_prefix: 'Cleanup Security Checks', test_array: @@studies_to_clean)
+  end
+
+  teardown do
+    Rails.env = "test" # reset this value to prevent other env checks failing later as this will persist across all runs
+  end
+
+  test 'should validate hostname' do
+    assert StudyCleanupTools.permit_hostname?
+  end
+
+  test 'should validate billing project' do
+    billing_project = FireCloudClient::PORTAL_NAMESPACE
+    assert StudyCleanupTools.permit_billing_project?(billing_project)
+
+    bad_project = "this-is-not-valid"
+    refute StudyCleanupTools.permit_billing_project?(bad_project)
+  end
+
+  test 'should validate environment' do
+    assert StudyCleanupTools.permit_environment?
+
+    # test allowing development environment
+    Rails.env = 'development'
+    assert StudyCleanupTools.permit_environment?(allow_dev_env: true)
+
+    # ensure production-like environments are disallowed
+    Rails.env = 'staging'
+    refute StudyCleanupTools.permit_environment?
+  end
+
+  test 'should validate continuous integration' do
+    assert StudyCleanupTools.is_continuous_integration?
+  end
+
+  test 'should match workspace creator for service account' do
+    study = Study.first
+    workspace = ApplicationController.firecloud_client.get_workspace(study.firecloud_project, study.firecloud_workspace)
+    assert StudyCleanupTools.service_account_created_workspace?(workspace)
+
+    # negative test with mock workspace JSON
+    not_owned_by_portal = {workspace: {createdBy: @user.email}}
+    refute StudyCleanupTools.service_account_created_workspace?(not_owned_by_portal)
+  end
+
+  test 'should raise error on validation failure' do
+
+    # positive tests - known goods should return nil as there is no declared return value
+    # no arguments
```

Comment:
Yeah, rewriting halt_on_validation_fail to take a block will let you test it much more independently, because you can just test 
```halt_on_validation_fail { false }``` ```halt_on_validation_fail { true }```
```halt_on_validation_fail { nil }```
etc...

Reply:
32c1bc3